### PR TITLE
Optionally pass Nixpkgs through entry functions

### PR DESCRIPTION
**Pull request by ehmry**
The "builtins.fetchTarball" function is not available in some restricted
environments such as build jobs of the latest Hydra. Make it possible to
pass nixpkgs and nixpkgsStaticLinux when evaluating default.nix and
release.nix to avoid importing them internally. This does not change the
result of evaluation if no parameters are passed.

---

I have a [Hydra (IPv6 only)](https://hydra.hq.c3d2.de/jobset/dhall-haskell/master) that I use to build configurations that require the latest Dhall prelude and thus the latest interpreter. My Hydra uses a version that is about a year ahead of the Dhall Hydra and it will not allow `builtins.fetchTarball`, which is used here to pin Nixpkgs revisions. This patch moves the pinning of Nixpkgs to `default.nix` and `release.nix` so that Hydra can fetch these versions by configuration and pass them to `release.nix`.

As far as I can tell this patch does not have an effect on the build artifacts.

### CODE REVIEW

File: nix/pinnedNixpkgs.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
+let
```
[the remark this answers is not in the document]
Comment (ehmry) [reply nix/pinnedNixpkgs.nix]: Yes, I added a fixup, but I can also squash.

**Comment by ehmry** [comment]
`shell.nix` too.

**Comment by ehmry** [comment]
I've reverted and squashed changes to `default.nix` and added fallbacks to `nix/shared.nix`. This way the parameter handling only needs to be done in `nix/shared.nix` and `release.nix`, rather than all the little shell and test files.

**Comment by ehmry** [comment]
Thanks! Both Hydras say its good.

**Comment by ehmry** [comment]
So, this patch does break some builds on the Dhall-lang Hydra because [Hydra is passing](https://hydra.dhall-lang.org/jobset/dhall-haskell/master#tabs-configuration) a `nixpkgs` pinned to `release-17.09` to `release.nix`. Up until this patch, this argument was ignored, but is now overriding the revision in `pinnedNixpkgs`.

The fix to restore the previous outputs is either to remove the `nixpkgs` input from the Hydra job configuration, or to update it to `https://github.com/NixOS/nixpkgs.git 1d4de0d552ae9aa66a5b8dee5fb0650a4372d148 `.

**Comment by Gabriel439** [comment]
@ehmry: I'm going to see if I can remove the `nixpkgs` Hydra input.  Just in case you are curious, the place where it is defined is here:

https://github.com/dhall-lang/dhall-lang/blob/b760b4183c8c589e3572a7f9e049d60e69ca53dd/nixops/jobsets.nix#L38-L44

The main thing I need to check first is whether or not that would disrupt other `dhall-*` projects which might depend on that `nixpkgs` input.

**Comment by Gabriel439** [comment]
@ehmry: Sorry for the delay, but I just removed the `nixpkgs` input and that fixes the evaluation error on this branch

**Comment by ehmry** [comment]
Yes, my Hydra is working fine now with master, but if I may ask, how did you figure out how to test GitHub PRs? I have declarative spec working, but can't find documentation on PRs.

**Comment by Gabriel439** [comment]
@ehmry: You can find the NixOps specification for all `*.dhall-lang.org` infrastructure here:

https://github.com/dhall-lang/dhall-lang/tree/master/nixops

The relevant bit is here:

https://github.com/dhall-lang/dhall-lang/blob/master/nixops/logical.nix#L15-L41

The basic way that declarative spec for pull requests work is that you have two files:

* A top-level `project.json` file which is responsible for specifying how to fetch various inputs that are fed to `./jobsets.nix`

  One of those inputs has type `githubpulls`, which then uses Hydra's built-in `GitHubPulls.pm` plugin to fetch pull requests given a GitHub owner/repository

  In the linked code it's actually stored as a `project.nix` file which is converted to JSON using `builtins.toJSON`

* A `jobsets.nix` expression which converts generates the list of jobsets

  In this case, the Nix expression takes the `pullRequestsJSON` input produced by `project.json` and uses Nix logic to transform that into the jobset specification.

Then for each of the projects that use a declarative specification I configure them like this:

Screenshot_20191020_164552

For example, this is the change I made to remove the `nixpkgs` input from each declarative project:

https://github.com/dhall-lang/dhall-lang/pull/782

**Comment by ehmry** [comment]
Thanks! I have this figured out now.

**Comment by Gabriel439** [comment]
@ehmry: You're welcome! :slightly_smiling_face:

**Pull request merged**